### DISCUSSION TITLE
No data returned from useRouteData hook

### DISCUSSION THREAD

**Discussion by TGPSKI** [react-static/react-static#1480] [General] [answered]
Hi everyone,

I'm an experienced react-static developer with two sites under my belt. I'm attempting to use the latest release to build another page, but I am struggling with the `useRouteData` hook. I cannot get any route data.

My static config sources markdown files from `content/collections/{template_dir}`. Logging the returned value of `testimonials`, I see the expected content array. For whatever reason, this data is not being propagated to the route.

static.config.js
```
const routes = async () => {
  const { testimonials } = await jdown('content', {
    parseMd: false,
    fileInfo: true,
    assets: { output: './public' }
  });

  return [
    {
      path: '/',
      template: './src/App.js',
    },
    {
      path: '404',
      template: './src/App/404.js',
    },
    {
      path: '/contact',
      template: './src/App/Contact.js'
    },
    {
      path: '/testimonials',
      template: './src/App/Testimonials.js',
      getData: () => ({ testimonials })
    }
  ];
};
````

In my application root file, I am using `reach-router` to provide the router component, and <Root> is sourced from `react-static`.

App.js
```
const App = () => {
  return (
    <Root>
      <OuterCenteredContainer>
        <MaxWidthContainer>
          <StickyFooterFlex>
            <Header />
            <Router>
              <Home path="/" />
              <Contact path="/contact"/>
              <Testimonials path="/testimonials"/>
            </Router>
          </StickyFooterFlex>
          <Footer />
        </MaxWidthContainer>
      </OuterCenteredContainer>
    </Root>
  )
}
```

The page I'm looking to pass routeData to is `/testimonials`. Here, I'm rendering a very basic page which should just log the result of `props` and `testimonials` to the console. 

Testimonials.js
```
import React from 'react';
import { useRouteData } from 'react-static';

import { shuffle } from 'lodash';
import shortid from 'shortid';

const TestimonialsCard = ({}) => ({});

const Testimonials = props => {
  console.log(props);
  const {testimonials} = useRouteData();
  console.log(testimonials);
  return (
    <div>foo</div>
    )
}

export default Testimonials;
```

The output from these console.log statements are as follows:

![image](https://user-images.githubusercontent.com/12532803/88442342-fa04e700-cdc8-11ea-9483-ee6f067b4d1f.png)

What am I doing wrong here?

**Comment by SleeplessByte** [comment]
Hi there!

`useRouteData` only works inside `<Routes>` _I think_, because that component knows which route it's at (see: https://github.com/react-static/react-static/blob/master/docs/guides/dynamic-routes-react-router.md).

I'm not certain, but I _think_ that if you import the [`routePathContext`](https://github.com/react-static/react-static/blob/master/packages/react-static/src/browser/context/routePathContext.js) and supply the path via that (e.g. using the location), it should work.

```tsx
<Router>
  <YourCustomComponentThatSetsRoutePathContextCorrectly>
     <Home path="/" />
     <Contact path="/contact"/>
     <Testimonials path="/testimonials"/>
     <Routes />
  </YourCustomComponentThatSetsRoutePathContextCorrectly>
</Router>
```

**Reply by TGPSKI** [reply]
I'm going to give it a shot with react-router. I've tried putting non-route components inside the `<Router>` for reach, and it complains every time.

**Reply by SleeplessByte** [reply]
Did this work?

**Reply by TGPSKI** [reply]
Will update - haven't tried yet.

**Comment by TGPSKI** [comment]
Unfortunately, I still haven't been able to get this working. Migrating from @reach/router (well supported by react-static documentation) to react-router-dom (poorly supported by react static documentation) has been full of bugs and problems. I wish we could get working examples that show current usage in the monorepo - every example shows @reach/router, and the "guide" (https://github.com/react-static/react-static/blob/master/docs/guides/dynamic-routes-react-router.md) is not working at all for me.

**Reply by TGPSKI** [reply]
Here's just a snippet. I can change the URL, but components do not change. Is there a guide I am missing? 

```
import React from 'react';
import { Root, Routes } from 'react-static';
import { Switch, Route } from 'react-router-dom'

...

const App = () => {
  return (
    <Root>
      <OuterCenteredContainer>
        <MaxWidthContainer>
          <StickyFooterFlex>
            <Header />
            <Switch>
              <Route path="/" >
                <Home/>
              </Route>
              <Route path="/contact" >
                <Contact/>
              </Route>
              <Route path="/testimonials">
                <Testimonials/>
              </Route>
            </Switch>
          </StickyFooterFlex>
          <Footer />
        </MaxWidthContainer>
      </OuterCenteredContainer>
    </Root>
  )
}
```

**Reply by TGPSKI** [reply]
Ok, this one is specific to react-router, an `exact` prop is required for the root path, otherwise none of the routes will render.

It would make me so happy if the examples in this repository were up to date! None of the examples show react-router, even though its the only way to actually get route data using `useRouteData`.

**Comment by TGPSKI** [comment]
Does anyone have a working example showing usage of react-router-dom + useRouteData? I am pulling my hair out trying to get a basic function to work.

**Comment by TGPSKI** [comment]
I'm moving back and forth between reach and react routers trying to get a working solution, haven't gotten there yet. I'll keep updating my progress.

**Reply by SleeplessByte** [reply]
I'm assuming you can't get it to work in the `Home`, `Testimonials` or `Contact` route right? I would suggest creating those via `static.js` instead (so they get populated via `Routes`), where `useRouteData` does work.

**Comment by nathonius** [comment, marked as answer]
React-static provides route data _only_ to the routes that it renders based on your static.config.js. The data is stored in a context that is part of the `<Routes/>` component, so you can't access it outside of that.

If you want some data provided to dynamically routed components (any routes not created in static.config.js and not rendered by `<Routes/>`), you can use `getSiteData` and `useSiteData` instead. That data is provided to your entire application.
